storage/src/main/scala/coop/rchain/storage/regex/Fsm.scala

```diff
@@ -0,0 +1,757 @@
+    * This procedure uses arbitrary amounts of memory but is very fast. There
+    * may be more efficient ways to do this, that I haven't investigated yet.
+    * You can use this in list comprehensions.
+    */
+  def strings: Stream[String] = {
+    // Many FSMs have "dead states". Once you reach a dead state, you can no
+    // longer reach a final state. Since many strings may end up here, it's
+    // advantageous to constrain our search to live states only.
+    val livestates = states.filter(state => isLive(state))
+
+    // We store a list of tuples. Each tuple consists of an input string and the
+    // state that this input string leads to. This means we don't have to run the
+    // state machine from the very beginning every time we want to check a new
+    // string.
+    val pending = mutable.Queue[(String, Int)]()
+    val pendingFinals = mutable.Queue[String]()
+    var needInitial = true
+
+    def nextStr: Option[String] =
+      if (pendingFinals.nonEmpty) {
+        Some(pendingFinals.dequeue())
+      } else {
+        //Initial entry (or possibly not, in which case this is a short one)
+        if (needInitial) {
+          needInitial = false
+          val cstate = this.initialState
+          if (livestates.contains(cstate)) {
+            pending += "" -> cstate
+            if (finalStates.contains(cstate)) {
+              pendingFinals += ""
+            }
+          }
+        }
+
+        while (pendingFinals.isEmpty && pending.nonEmpty) {
+          val (pendingString, pendingState) = pending.dequeue()
+
+          transitions.getOrElse(pendingState, Nil).foreach({
+            case (nextSymbol, nextState) => {
+              val nextString = pendingString + nextSymbol
+              if (livestates.contains(nextState)) {
+                pending += nextString -> nextState
+                if (finalStates.contains(nextState)) {
+                  pendingFinals += nextString
+                }
+              }
+            }
+          })
+        }
+
+        if (pendingFinals.nonEmpty) {
+          Some(pendingFinals.dequeue())
+        } else {
+          None
+        }
+      }
+
+    def genStream: Stream[String] = {
+      val str = nextStr
+      if (str.isDefined) {
+        str.get #:: genStream
+      } else {
+        Stream.empty[String]
+      }
+    }
+    genStream
+  }
+
+  /**
+    * Given an FSM and a multiplier, return the multiplied FSM.
+    */
+  def times(multiplier: Int): Fsm = {
+    require(multiplier >= 0, "Can't multiply an FSM by " + multiplier)
+    //here we always work with "currentState -> iteration" pairs
+    def initial = mutable.Set(this.initialState -> 0)
+
+    def follow(crawlState: mutable.Set[(Int, Int)],
+               symbol: Char): Option[mutable.Set[(Int, Int)]] = {
+      val next = mutable.Set[(Int, Int)]()
+
+      crawlState.foreach { case (fsmState, iteration) =>
+        if (iteration < multiplier) {
+          val transition = transitions.get(fsmState)
```

Comment:
Maybe could also be simplified by a for expression.

Reply:
Simplified as a part of get.get.get... elimination